crates/mdns/src/lib.rs

```diff
@@ -1,172 +1,435 @@
+//!     println!("mDNS neighbourhood : {:#?}", &mdns.records());
+//!
+//!     thread::sleep(Duration::from_millis(100));
+//! }
+//! ```
 
 #![feature(try_trait)]
+#![feature(never_type)]
+#![feature(drain_filter)]
 
-extern crate byteorder;
-extern crate net2;
+use log::{debug, error, trace};
+// Used to clean our buffer to avoid mixing messages together.
+use url::Url;
+use zeroize::Zeroize;
+
+use std::{
+    net::{self, SocketAddr, ToSocketAddrs},
+    time::Instant,
+};
+
+use lib3h_discovery::{error::DiscoveryResult, Discovery};
+
+pub mod error;
+pub use error::{MulticastDnsError, MulticastDnsResult};
+
+pub mod dns;
+pub use dns::*;
+
+pub mod builder;
+pub use builder::MulticastDnsBuilder;
+
+pub mod record;
+use record::{MapRecord, Record};
 
 // 20 byte IP header would mean 65_507... but funky configs can increase that
 // const READ_BUF_SIZE: usize = 60_000;
 // however... we don't want to accept any packets that big...
 // let's stick with one common block size
 const READ_BUF_SIZE: usize = 4_096;
 
-#[cfg(not(target_os = "windows"))]
-use net2::unix::UnixUdpBuilderExt;
+/// Delay between probe query, 250ms by default.
+const _PROBE_QUERY_DELAY_MS: u64 = 250;
 
-use std::net::ToSocketAddrs;
+/// Listening port of this mDNS service.
+const SERVICE_LISTENER_PORT: u16 = 8585;
 
-pub mod error;
-pub use error::{MulticastDnsError, MulticastDnsResult};
+/// Threshold value used to getting ourselves out of a potential
+/// infinite loop during probe
+const _FAIL_SAFE_TRESHOLD: u16 = 1_000;
 
-pub mod dns;
-pub use dns::*;
+/// mDNS multicast IPv4 address.
+const MDNS_MULCAST_IPV4_ADRESS: &str = "224.0.0.251";
 
-/// mdns configuration
-#[derive(Clone, Debug)]
-pub struct Config {
-    pub bind_address: String,
-    pub bind_port: u16,
-    pub multicast_loop: bool,
-    pub multicast_ttl: u32,
-    pub multicast_address: String,
-}
+/// Default bind adress.
+const DEFAULT_BIND_ADRESS: &str = "0.0.0.0";
+
+/// Default "time to live" value for a new record.
+const DEFAULT_TTL: u32 = 255;
 
-/// mdns builder
-pub struct Builder {
-    config: Config,
+/// Default amount of time between two queries.
+const DEFAULT_EVERY_MS: u128 = 30_000;
+
+/// an mdns instance that can send and receive dns packets on LAN UDP multicast
+pub struct MulticastDns {
+    /// Our IP address bound to UDP Socket, default to `0.0.0.0`
+    pub(crate) bind_address: String,
+    /// Port used by the mDNS protocol. mDNS use the `5353` by default
+    pub(crate) bind_port: u16,
+    /// If true, multicast packets will be looped back to the local socket
+    pub(crate) multicast_loop: bool,
+    /// Time to Live: default to `255`
+    pub(crate) multicast_ttl: u32,
+    /// Multicast address used by the mDNS protocol: `224.0.0.251`
+    pub(crate) multicast_address: String,
+    /// Determine if we need to query / announce.
+    pub(crate) timestamp: Instant,
+    /// The amount of time we should wait between two queries.
+    pub(crate) every: u128,
```

Comment:
let's put a unit on here so it's immediately known. If it is milliseconds, we shouldn't need a u128, a u64 should hold it just fine:

```rust
pub(crate) query_interval_ms: u64,
```